tests/test_output.py

```diff
@@ -0,0 +1,511 @@
+
+    with patch("builtins.open", mock_open()) as mock_file:
+        output.log_line(msg)
+
+        mock_file.assert_called_once_with("path/to/slips.log", "a")
+        handle = mock_file()
+        handle.write.assert_called_once_with(expected_log_content)
+
+
+def test_print_no_pbar():
+    """Test printing when has_pbar is False."""
+    output = ModuleFactory().create_output_obj()
+    output.has_pbar = False
+    output.tell_pbar = MagicMock()
+    sender = "SenderName"
+    txt = "This is a test message."
+
+    with patch("builtins.print") as mock_print:
+        output.print(sender, txt)
+
+    (mock_print.assert_called_once_with(f"[{sender}] {txt}", end="\n"))
+    output.tell_pbar.assert_not_called()
+
+
+def test_print_pbar_finished():
+    """Test printing when pbar is finished."""
+    output = ModuleFactory().create_output_obj()
+    output.has_pbar = True
+    output.pbar_finished = MagicMock()
+    output.pbar_finished.is_set.return_value = True
+    output.tell_pbar = MagicMock()
+    sender = "SenderName"
+    txt = "This is a test message."
+
+    with patch("builtins.print") as mock_print:
+        output.print(sender, txt)
+
+    (mock_print.assert_called_once_with(f"[{sender}] {txt}", end="\n"))
+    output.tell_pbar.assert_not_called()
+
+
+def test_print_pbar_active_with_sender():
+    """Test printing with active pbar and a sender."""
+    output = ModuleFactory().create_output_obj()
+    output.has_pbar = True
+    output.pbar_finished = MagicMock()
+    output.pbar_finished.is_set.return_value = False
+    output.tell_pbar = MagicMock()
+    sender = "SenderName"
+    txt = "This is a test message."
+
+    with patch("builtins.print") as mock_print:
+        output.print(sender, txt)
+
+    (
+        output.tell_pbar.assert_called_once_with(
+            {"event": "print", "txt": f"[{sender}] {txt}"}
+        )
+    )
+    mock_print.assert_not_called()
+
+
+def test_print_pbar_active_no_sender():
+    """Test printing with active pbar and no sender."""
+    output = ModuleFactory().create_output_obj()
+    output.has_pbar = True
+    output.pbar_finished = MagicMock()
+    output.pbar_finished.is_set.return_value = False
+    output.tell_pbar = MagicMock()
+    sender = ""
+    txt = "This is a message with no sender."
+
+    with patch("builtins.print") as mock_print:
+        output.print(sender, txt)
+
+    (output.tell_pbar.assert_called_once_with({"event": "print", "txt": txt}))
+    mock_print.assert_not_called()
+
+
+def test_handle_printing_stats_pbar_not_finished():
+    """Test when pbar is not finished, stats should be sent to pbar."""
+    output = ModuleFactory().create_output_obj()
+    output.has_pbar = True
+    output.pbar_finished = MagicMock()
+    output.pbar_finished.is_set.return_value = False
+    output.pbar_sender_pipe = MagicMock()
+    stats = "Analyzed IPs: 10"
+
+    output.handle_printing_stats(stats)
+
+    (
+        output.pbar_sender_pipe.send.assert_called_once_with(
```

Comment:
here we shouldnt be checking output.pbar_sender_pipe.send
pbar_sender_pipe.send() is not in the code of handle_printing_stats(), it is in the code of tell_pbar()
so we should be check tell_pbar() instead